Introduce a release issue template (#5826)
filecoin-project/lotus: documentation/misc/RELEASE_ISSUE_TEMPLATE.md

> Release Issue Template

# Lotus X.Y.Z Release

We're happy to announce Lotus X.Y.Z...

> ## 🗺 What's left for release

Reviewer: Is this a list of issues that **must** be completed for this release? if so, can we restrict that no more tickets can be added X days before the `Estimated shipping date`, for which i would suggest 7 for X(around when the rc is suppose to be cut)

Author: Yeah, I would honestly like to get rid of this section. Thoughts?

I don't want "must be completeds" except in extraordinary situations.

Reviewer 2: "must dos" may be common for state upgrade releases which have an agreed-upon FIP across all implementations. Imagine that should be included here

> <List of items with PRs and/or Issues to be considered for this release>

Reviewer 3: @jennijuju is this going to be a link to a notion?

Reviewer: I think it should be a list of issues in this section.

Author: Note that this should ideally be very few items for minor releases -- I really wanna move away from blocking releases on stuff that's "almost ready".

## 🚢 Estimated shipping date

<Date this release will ship on if everything goes to plan (week beginning...)>

## 🔦 Highlights

< top highlights for this release notes >

## Changelog

> < changelog generated by bin/mkreleaselog >

Reviewer 3: I'd actually consider not including this in the issue. It tends to make it a bit unwieldy.

Reviewer 3: But up to you.

Author: sgtm

## ✅ Release Checklist

First steps:

>   - [ ] Fork a new branch (`release/vX.Y.Z`) from `master` and make any further release related changes to this branch. If any "non-trivial" changes get added to the release, uncheck all the checkboxes and return to this stage.

Reviewer: ```suggestion
  - [ ] Right after the previous release is tagged, fork a new branch (`release/vX.Y.Z`) from `master` and make any further release related changes to this branch. If any "non-trivial" changes get added to the release, uncheck all the checkboxes and return to this stage.
```

Author: No, that's not quite right, I don't think?

Reviewer 2: do you want to modify the process here for your case? https://blog.ipfs.io/2019-08-14-ipfs-release-process/

>   - [ ] Fork a new branch (`release/vX.Y.Z`) from `master` and make any further release related changes to this branch. If any "non-trivial" changes get added to the release, uncheck all the checkboxes and return to this stage.

Author: Quite right!

>   - [ ] Prep the changelog

Reviewer: +1 to only add the changelog in `CHANGELOG.md`

>   - [ ] Prep the changelog

Author: Clarifying!

>   - [ ] Bump the version in `version.go` in the `master` branch to `vX.(Y+1).0-dev`.

Author: Agreed, @jennijuju, making that change

>   - [ ] Bump the version in `version.go` in the `master` branch to `vX.(Y+1).0-dev`.

Author: Synced with @jennijuju offline, think it's fine the way it is.

>   - [ ] Follow the RC release process to cut the first RC.

Author: Yeah, it's redundant (it does mean "do the next thing in this list").

>   - [ ] Follow the RC release process to cut the first RC.

Author: Removing.

    
Prepping an RC:

- [ ] version string in `build/version.go` has been updated (in the `release/vX.Y.Z` branch).

> - [ ] tag commit with `vX.Y.Z-rcN`

Reviewer 4: Fine for now, but these seem like steps that can be scripted in future.

Author: Noted in a next steps doc.

> - [ ] cut a pre-release on [github](https://github.com/filecoin-project/lotus/releases)

Reviewer 4: Can we provide a URL that prepulates the form?  At the minimum we can do https://github.com/filecoin-project/lotus/releases/new?prerelease=true 

Other fields that can be filled in https://github.com/isaacs/github/issues/1410#issuecomment-442240267

What should be specified for the other values in the form?

This is something a script could help do (e.g., by taking in X/Y/Z it can generate more of the URL parameters).

Author: This is neat! I grabbed the prerelase param. I think everything else will need to be specific to the version, making a note for next steps.

Testing an RC:

- [ ] **Stage 0 - Automated Testing**
  - Automated Testing
    - [ ] CI: Ensure that all tests are passing.

>     - [ ] Testground tests

Reviewer 4: Do these automatically run?  If not, where does someone look to confirm this?

Reviewer 3: They do. @coryschwartz what's the status on getting the testground test results back into lotus?

Reviewer 5: I have a couple of open PRs for working with TG. The main one is https://github.com/filecoin-project/lotus/pull/5887, which just adds testground tests at the bottom of PRs. The other PRs are a workflow for running more or longer-running testground tests that get triggered just during release...like on a release issue, or a "chatops" interface in which specific tests can be asked for in issues or PRs and their results included in github discussions. 
```
https://github.com/filecoin-project/lotus/pull/5888
https://github.com/filecoin-project/lotus/pull/5887
https://github.com/filecoin-project/lotus/pull/5916
```

- [ ] **Stage 1 - Internal Testing**

>   - Upgrade our testnet infra

Reviewer 4: What things do we need to validate before we can move on?  Is there an alarm that we should check to make sure isn't firing?

Author: Mostly that nodes stay in sync, I think. Adding that.

    - [ ] 1 bootstrap node
    - [ ] 1 miner
    - [ ] Scratch nodes
    - [ ] Wait 24 hours

>     - [ ] Remaining testnet infra

Reviewer 4: How much do we need to wait after this?

Author: I don't think we do -- upgrading the remainder of testnet infra and the small subset of mainnet infra can be done together.

  -  Upgrade our mainnet infra

>     - [ ] Subset of development full archival nodes

Reviewer 4: I assume this gets checked when we have deployed to a subset of full archival nodes.  

Will we also have a checkbox once we have validated that everything is good?

Author: Added one for the simple "are nodes still in sync" question, but there's more that needs to be done.

>     - [ ]Subset of bootstrappers (1 per region)

Reviewer 4: Add space after ]

Author: 👀

>     - Report on new block validation time

Reviewer 4: Where do we report this?  
Who is going to decide whether this is ok or not?

Author: I think this is TBD, it's the first item on my list of "what's next?"

>     - TODO: What other stats would we care about?

Author: @magik6k @Stebalien thoughts?

Reviewer 6: Memory/CPU use are worth following

Reviewer 4: Can we link to the dashoard we're going to look at?

Reviewer 3: Number of goroutines. IPLD block read latency. Bandwidth usage (by protocol, ideally).

>     - If anything has worsened significantly, investigate + fix

Reviewer 4: Can we be a bit more prescriptive here?  Maybe: generate a graph comparing current memory/CPU (post deployment) to past memory/CPU (pre deployment).  Confirm no negative change.  Post a snapshot of the graph in X.

>   - Confirm the following work (some combination of Testground / Calibnet / Mainnet / MinerX)

Reviewer: ```suggestion
  - Confirm the following work (some combination of Testground / Calibnet / Mainnet / lotus early testers)
```
```suggestion
  - Confirm the following work (some combination of Testground / Calibnet / Mainnet / MinerX)
```
```suggestion
  - Confirm the following work (some combination of Testground / Calibnet / Mainnet / MinerX)
```

Reviewer 7: Let's call them "Beta Users" rather than "lotus early testers" or "MinerX".  The system has gone through some testing by the time these communities are engaged.

    - [ ] Seal a sector
    - [ ] make a deal
    - [ ] Submit a PoSt

>     - [ ] (ideally) let a sector go faulty, and see it be recovered

Reviewer 4: Can we remove the ideally qualifier?  If this is important, can we commit to doing it?

Author: I'm gonna call it optional. There's gonna be select situations where it's important to look at this, and it's not the easiest thing to arrange. We can write some shed tool that says "let my next set of sectors go faulty", but it's still 24 hours before it'll actually be confirmed.

I'm concerned about how "heavy" this checklist is -- we can add more later.

    
- [ ] **Stage 2 - Community Dev Testing**

>   - [ ] Inform MinerX / early testers

Reviewer 4: How does someone do this?

Do we wait some amount of time or for some number of them to give the thumbs up before we move forward?

Author: - @jennijuju messages them in Slack
- I think that's an open question I wanna descope from the initial version of this.

>   - [ ] Inform MinerX / early testers

Reviewer: ```suggestion
  - [ ] early miner testers
```

>   - [ ] Inform MinerX / early testers

Reviewer: ```suggestion
  - early miner testers (@lotus-early-testers-miner in Filecoin Slack)
```

>   - [ ] Ask close ecosystem partners to test their projects with the upgrade

Reviewer 4: How do we ask?  Do we have a template we send?

Do we give them a timeline of when we'd like to receive a response by before we move forward?  PHrased another way, should we say "We're planning to move forward with this release unless we hear otherwise from you by YYYY-MM-DD."

    - [ ] Powergate

>     - TODO: List of partners

Author: Have asked for suggestions, can be completed later

>   - [ ] Ask close ecosystem partners to test their projects with the upgrade
>     - [ ] Powergate
>     - TODO: List of partners

Reviewer: ```suggestion
  - [ ] Ask close ecosystem partners to test their projects with the upgrade (@lotus-early-testers-eco-dev)
```

- [ ] **Stage 3 - Community Prod Testing**
  - [ ] Documentation
    - [ ] Ensure that [CHANGELOG.md](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) is up to date

>     - [ ] TODO: Other docs checks?

Author: @jennijuju any ideas for elaboration here?

Reviewer 6: Would add config docs - https://docs.filecoin.io/get-started/lotus/configuration-and-advanced-usage/#configuration

Reviewer 4: This is interesting.  Is it possible that folks are adding new functionality but in the PR not adding the relevant doc updates?  If so, is it a manual audit to go through commits looking for this?  That seems rough.

Author: @BigLep So this is tricky. Several thoughts:

- Yes, PRs often land without necessary docs updates.
- Most documentation lives outside of the Lotus repo.
- We haven't really established a workflow for doc updates -- it's somewhat haphazard right now.

  - [ ] Invite the wider community through (link to the release issue):

>     - [ ] TODO: How should we announce this?

Author: whatever @jennijuju wants to do

    
- [ ] **Stage 4 - Release**
  - [ ] Final preparation
    - [ ] Verify that version string in [`version.go`](https://github.com/ipfs/go-ipfs/tree/master/version.go) has been updated.
    - [ ] Ensure that [CHANGELOG.md](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) is up to date
    - [ ] Ensure that [README.md](https://github.com/filecoin-project/lotus/blob/master/README.md) is up to date
    - [ ] Merge `release-vX.Y.Z` into the `releases` branch.
    - [ ] Tag this merge commit (on the `releases` branch) with `vX.Y.Z`.

>     - [ ] Cut the release on Github.

Reviewer 4: Similar to above, lets prepoulate the url as much as possible.

Author: ack

    - [ ] Final announcements
        - [ ] Update network.filecoin.io

>         - [ ] TODO: What / where else?

Author: @jennijuju

>         - [ ] TODO: What / where else?

Reviewer: ```suggestion
        - [ ] Add a comment when the final release is tagged, example [here](https://github.com/filecoin-project/lotus/discussions/5905#discussioncomment-571752)
        - [ ] repost in #fil-lotus in filecoin slack
```

Reviewer: I wonder if this is something we can eventually have community member to do it, ambassdors maybe?

- [ ] **Post-Release**
  - [ ] Merge the `releases` branch back into `master`, ignoring the changes to `version.go` (keep the `-dev` version from master).

>   - [ ] Create an issue using this release issue template for the _next_ release.

Reviewer 4: I think we can prepopulate this URL as well.

> ## ❤️ Contributors

Reviewer 4: Is the point of this section to pull out who contributed?

Author: Yup, example from IPFS
![Screen Shot 2021-04-08 at 12 59 32 AM](https://user-images.githubusercontent.com/6530144/113970871-af67b580-9805-11eb-85b1-86a8a517eaf6.png)

< list generated by bin/mkreleaselog >

Would you like to contribute to Lotus and don't know how? Well, there are a few places you can get started:

> - TODO

Author: write this and the next one when i'm not lazy...or merge as is and complete later

Reviewer 2: https://github.com/filecoin-project/lotus#contribute
https://github.com/filecoin-project/lotus/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22
https://github.com/filecoin-project/FIPs
https://spec.filecoin.io/#section-systems.filecoin_nodes

## ⁉️ Do you have questions?

> TODO

Reviewer: Leave a comment [here](link to release discussion) if you have any questions.

> Release Issue Template

# Lotus X.Y.Z Release

We're happy to announce Lotus X.Y.Z...

> ## 🗺 What's left for release

Reviewer: ```suggestion
## 🗺 Must-dos for the release

## 🌟 Nice-to-haves for the release
```

    - [ ] Create a lotus disucssion, example [here](https://github.com/filecoin-project/lotus/discussions/5595)
    - [ ] Link the disucssion in #fil-lotus on Filecoin slack
    
- [ ] **Stage 4 - Release**
  - [ ] Final preparation
    - [ ] Verify that version string in [`version.go`](https://github.com/ipfs/go-ipfs/tree/master/version.go) has been updated.
    - [ ] Ensure that [CHANGELOG.md](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) is up to date
    - [ ] Ensure that [README.md](https://github.com/filecoin-project/lotus/blob/master/README.md) is up to date
    - [ ] Merge `release-vX.Y.Z` into the `releases` branch.
    - [ ] Tag this merge commit (on the `releases` branch) with `vX.Y.Z`.
    - [ ] Cut the release [here](https://github.com/filecoin-project/lotus/releases/new?prerelease=true&target=releases).
    - [ ] Final announcements
        - [ ] Update network.filecoin.io
        - [ ] Add a comment when the final release is tagged, example [here](https://github.com/filecoin-project/lotus/discussions/5905#discussioncomment-571752)
        - [ ] repost in #fil-lotus in filecoin slack

- [ ] **Post-Release**

>   - [ ] Merge the `releases` branch back into `master`, ignoring the changes to `version.go` (keep the `-dev` version from master).

Reviewer: Update http://network.filecoin.info/, guides [here](https://docs.google.com/document/d/1AkR4LMO-lyintgVxrCMeorutiE-IW4DGKvit_f6lgSY/edit)

    - [ ] Check if any [config](https://docs.filecoin.io/get-started/lotus/configuration-and-advanced-usage/#configuration) updates are needed
  - [ ] Invite the wider community through (link to the release issue):
    - [ ] Create a lotus disucssion, example [here](https://github.com/filecoin-project/lotus/discussions/5595)
    - [ ] Link the disucssion in #fil-lotus on Filecoin slack
    
- [ ] **Stage 4 - Release**
  - [ ] Final preparation
    - [ ] Verify that version string in [`version.go`](https://github.com/ipfs/go-ipfs/tree/master/version.go) has been updated.
    - [ ] Ensure that [CHANGELOG.md](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) is up to date
    - [ ] Ensure that [README.md](https://github.com/filecoin-project/lotus/blob/master/README.md) is up to date
    - [ ] Merge `release-vX.Y.Z` into the `releases` branch.
    - [ ] Tag this merge commit (on the `releases` branch) with `vX.Y.Z`.
    - [ ] Cut the release [here](https://github.com/filecoin-project/lotus/releases/new?prerelease=true&target=releases).
    - [ ] Final announcements

>         - [ ] Update network.filecoin.io

Reviewer: ```suggestion
        - [ ] Update network.filecoin.io for mainnet, calib and nerpa.
```

        - [ ] Add a comment when the final release is tagged, example [here](https://github.com/filecoin-project/lotus/discussions/5905#discussioncomment-571752)

>         - [ ] repost in #fil-lotus in filecoin slack

Reviewer: ```suggestion
        - [ ] repost in #fil-lotus in filecoin slack
        - [ ] Inform node provides (Protofire, Digital Ocean..)
```
